added BYOS ("Bring Your Own SERVICE_ACCOUNT") for GCS authentication

akashShanmugraj opened the pull request:
#### Description

This pull request allows the application to authenticate its Google Cloud Storage buckets through given SERVICE_ACCOUNT credential. The path to JSON file containing the service account credentials must be pre-set to the environment variable `GCP_CREDENTIAL_PATH`

#### Additional information...

- [] Have you run the test suite?
  Many changes need to be tested. Please run the test suite
  and include the output of ```cd test && ./problems```.

I ran `cargo test`.  All tests were successful.  `cargo build` was also successful. 
But I was not able to verify if it actually works

akashShanmugraj committed:
Add GCP credential path to server configuration

akashShanmugraj commented:
The `./taskchampion-sync-server` executable spins up an Actix Webserver and I am not sure which routes I should visit to test it with my account

djmitche commented:
> The `./taskchampion-sync-server` executable spins up an Actix Webserver and I am not sure which routes I should visit to test it with my account

Don't worry about that -- it's not relevant for this project.

djmitche commented on taskchampion/lib/src/lib.rs:
```diff
@@ -129,7 +129,6 @@ mod util;
 pub mod annotation;
 pub use annotation::*;
 pub mod atomic;
-pub use atomic::*;
```
Can you make a separate PR for this?

akashShanmugraj replied:
Okay!

djmitche commented on taskchampion/taskchampion/src/server/cloud/gcp.rs:
```diff
@@ -241,13 +243,21 @@ mod tests {
     fn make_service() -> Option<(GcpService, impl Fn(&str) -> Vec<u8>)> {
-        let Ok(bucket) = std::env::var("GCP_TEST_BUCKET") else {
-            return None;
+        let bucket = match std::env::var("GCP_TEST_BUCKET") {
+            Ok(val) => val,
+            Err(_) => return None,
+        };
```
Why did this change from the `let Ok(..) = .. else { .. }` syntax?

akashShanmugraj replied:
Ah I thought this was a bit better in terms of exception handling, I'll change it back immediately

djmitche reviewed (changes requested):
This looks good! I think it needs three more things:

1. Documentation update in `doc/man/task-sync.5.in`
2. Make the credential path optional. In Rust, this should be `Option<..>`, and that means the FFI needs to use TCStringType::Null to represent the `None` variant.
3. Add support for supplying this parameter from `taskrc`. If you look for `sync.gcp.bucket`, you should find all of the places that need to be updated.

akashShanmugraj committed:
Refactor GCP service initialization in tests

akashShanmugraj committed:
Made credential path optional in TCServer Initialization

akashShanmugraj commented:
> 2\. Make the credential path optional. In Rust, this should be `Option<..>`,

I tried to add the optional modifier to `credentialpath` but I am not sure how to fix this warning:

<img width="1119" alt="Screenshot 2024-01-21 at 6 01 00 AM" src="https://github.com/djmitche/taskwarrior/assets/65720968/5dc2b21d-fd8a-4bc1-8ff1-aed1b8735054">

Plus I hope my string conversion here is correct for None too?
```rust
let credentialpath = match credentialpath {
                Some(path) => unsafe { TCString::val_from_arg(path) }.into_string()?,
                None => String::new(), // or provide a default path
            };
```

djmitche commented:
You can't pass an `Optional<..>` value from C, so you'll need to keep `tc_server_new_gcp` taking a TCString argument. However, TCString has an `is_null` method which you can use to convert it to either `Some(String)` or `None`. Something like
```
let credentialpath = if credentialpath.is_null() {
  None
} else {
  // SAFETY: ..
  Some(unsafe { TCString::val_from_arg(path) }.into_string()?)
}
```

akashShanmugraj commented:
Am I handling the `Option<String>` correctly?

```rust
let credentialpath = if credentialpath.is_null() {
                None
            } else {
                Some(unsafe { TCString::val_from_arg(credentialpath) }.into_string()?)
            };

            let credentialpath = credentialpath.unwrap();
```

akashShanmugraj commented:
> 3\. If you look for `sync.gcp.bucket`, you should find all of the places that need to be updated.

Where should I find the `sync.gcp.bucket`? It is a file?

djmitche commented:
> > 3. If you look for `sync.gcp.bucket`, you should find all of the places that need to be updated.
> 
> Where should I find the `sync.gcp.bucket`? It is a file?

It's an existing configuration parameter that is handled similarly to the new one.

Try
```
git grep sync.gcp.bucket
```
to find where it is referenced.

djmitche commented:
> Am I handling the `Option<String>` correctly?

It's hard to say without seeing the patch. The compiler will probably tell you if you're not :)

> ```rust
>             let credentialpath = credentialpath.unwrap();
> ```

This doesn't seem right -- the statement before it carefully constructs an Option<String>, so it's odd to immediately unwrap that. It will panic in the `None` case!

akashShanmugraj commented:
But if I dont unwrap it, I am getting mismatched type error

<img width="901" alt="Screenshot 2024-01-21 at 7 21 36 AM" src="https://github.com/djmitche/taskwarrior/assets/65720968/465f825f-1434-4537-9fe7-c276b763e3f6">

I am not sure how to handle it

akashShanmugraj commented:
> It's an existing configuration parameter that is handled similarly to the new one.

Oh okay 
Now I should create a new parameter say `sync.gcp.credentials` at all the places where existing parameter `sync.gcp.bucket` is already defined, add and define that configuration everywhere.

akashShanmugraj committed:
Added documentation for GCP Sync

akashShanmugraj commented:
Maybe something like this then?
```rust
            let credpath:String;

            if credentialpath.is_null() {
                credpath = String::from("");
            } else {
                credpath = unsafe { TCString::val_from_arg(credentialpath) }.into_string()?;
            }
```

akashShanmugraj committed:
Map null credentialpath to default auth

akashShanmugraj committed:
added credentialpath to config

akashShanmugraj committed:
added credentialpath parameter to internal functions

djmitche commented on src/Context.cpp:
```diff
@@ -283,6 +283,7 @@ std::string configurationDefaults =
   "#sync.server.encryption_secret                 # Encryption secret for sync to a server\n"
   "#sync.server.origin                            # Origin of the sync server\n"
   "#sync.local.server_dir                         # Directory for local sync\n"
+  "#sync.gcp.credentialpath                       # Path to JSON file containing credentials to authenticate GCP Sync\n"
```
The configuration parameters have `_` between words, so this should be `credential_path`.

akashShanmugraj replied:
Okay!

akashShanmugraj replied:
Okay! 
I have changed parameters throughout, just in case

djmitche commented on doc/man/task-sync.5.in:
```diff
@@ -85,6 +85,42 @@ Then configure Taskwarrior with:
+ Select the role you just created and complete the service account creation process.
+
+ Now, in the Service Account dashboard, click into the new service account and select "keys" on the top menu bar.
+ Click on "ADD KEY" to create a new key or upload an existing key (a JSON key).
```
How would uploading an existing key work here? Maybe "or upload an existing key" should be omitted here?

akashShanmugraj replied:
In the Keys tab of every service account, there is an option to upload our own keys below the create new key option 

<img width="846" alt="Screenshot 2024-01-22 at 6 48 45 AM" src="https://github.com/djmitche/taskwarrior/assets/65720968/5d8da311-897a-47e9-92d8-cf47e9414595">

The user can upload his/her existing keys from the local storage

<img width="685" alt="Screenshot 2024-01-22 at 6 48 51 AM" src="https://github.com/djmitche/taskwarrior/assets/65720968/da5b5c53-c426-4ab1-a225-a2fb4c4cf176">

djmitche replied:
My question was, would someone who needs these instructions to know what to do be interested in uploading a cusotm X509 key? It seems unlikely, and I suspect anyone who did know how to generate an X509 key would not be reading these instructions.

akashShanmugraj replied:
Yes, seems right. I'll change it to _Download/Create a new key_.
Plus, if someone is so experienced that they must upload a custom X509 key, then they must definitely know how to use it here!

djmitche commented on taskchampion/lib/src/server.rs:
```diff
@@ -170,6 +170,7 @@ pub unsafe extern "C" fn tc_server_new_sync(
 #[no_mangle]
 pub unsafe extern "C" fn tc_server_new_gcp(
     bucket: TCString,
+    credentialpath: Option<TCString>,
```
Since this function is called from C++, it can't use Rust-specific types like `Option`. This will need to be just a plain `TCString`.

akashShanmugraj replied:
Ah I forgot to push the new changes, what do you think of this?
```rust
#[no_mangle]
pub unsafe extern "C" fn tc_server_new_gcp( 
    bucket: TCString,
    credential_path_argument: TCString,
    encryption_secret: TCString,
    error_out: *mut TCString,
) -> *mut TCServer {
    wrap(
        || {
            // SAFETY:
            //  - bucket is valid (promised by caller)
            //  - bucket ownership is transferred to this function
            let bucket = unsafe { TCString::val_from_arg(bucket) }.into_string()?;

            // SAFETY:
            //  - credential_path is valid (promised by caller)
            //  - credential_path ownership is transferred to this function
            // TODOAK003: push changes after confirming with @djmitche
            let credential_path:String;

            if credential_path_argument.is_null() {
                credential_path = String::from("");
            } else {
                credential_path = unsafe { TCString::val_from_arg(credential_path_argument) }.into_string()?;
            }

            // SAFETY:  
 
```

djmitche replied:
Close, but `credential_path` should be an `Option<String>`.

akashShanmugraj replied:
Okay!

djmitche commented on taskchampion/lib/src/server.rs:
```diff
@@ -180,6 +181,13 @@ pub unsafe extern "C" fn tc_server_new_gcp(
+            // SAFETY:
+            //  - credentialpath is valid (promised by caller)
+            //  - credentialpath ownership is transferred to this function
+            let credentialpath = match credentialpath {
```
Here I think you'll want to test whether `credentialpath == ""` and set the value passed to `ServerConfig::Gcp` to `None` for the empty string, and `Some(path)` otherwise.

akashShanmugraj replied:
I did something like this:

```rust
let credential_path:String;

if credential_path_argument.is_null() {
    credential_path = String::from("");
} else {
    credential_path = unsafe { TCString::val_from_arg(credential_path_argument) }.into_string()?;
}
```

And Gcp Serverconfig is initialised as it is already present, like
```rust
let server_config = ServerConfig::Gcp {
    bucket,
    credential_path: Some(credential_path),
    encryption_secret,
};
```

the reason I am passing `""` for the null case is because the credential_path in the authentication is checked against this empty string

```rust
let credentialpathstring = credential_path.clone().unwrap();
if credential_path.unwrap() == "" {
    config = rt.block_on(ClientConfig::default().with_auth())?;
} else {
    let credentials = rt.block_on(CredentialsFile::new_from_file(credentialpathstring))?;
    config = rt.block_on(ClientConfig::default().with_credentials(credentials))?;
}
```

djmitche commented on taskchampion/taskchampion/src/server/cloud/gcp.rs:
```diff
@@ -25,9 +26,17 @@ fn is_http_error<T>(query: u16, res: &std::result::Result<T, http::Error>) -> bo
 
 impl GcpService {
-    pub(in crate::server) fn new(bucket: String) -> Result<Self> {
+    pub(in crate::server) fn new(bucket: String, credentialpath:String) -> Result<Self> {
```
Since this is a pure Rust function, it should take the common Rust type, `Option<String>`, instead of special-casing the empty string. The conversion from the special-cased-empty-string to None is done in `tc_server_new_gcp`, which is a wrapper around `GcpService::new`.

akashShanmugraj replied:
So I need to change the datatype for both `bucket` and `credentialpath`?

djmitche replied:
No, Bucket can still be a String because it must be present.

akashShanmugraj replied:
okay 👍🏻

djmitche commented on taskchampion/taskchampion/src/server/cloud/gcp.rs:
```diff
@@ -242,12 +251,18 @@ mod tests {
         let Ok(bucket) = std::env::var("GCP_TEST_BUCKET") else {
-            return None;
-        };
+            return None;};
+        
+        let Ok(credentialpath) = std::env::var("GCP_CREDENTIAL_PATH") else {
+            return None;};
+            
```
Don't forget to run `cargo fmt` on your changes!

akashShanmugraj replied:
Okay!

djmitche reviewed (changes requested):
Looking good!

djmitche committed:
Add support for cloud sync, specifically GCP (#3223)

* Add support for cloud sync, specifically GCP

This adds generic support for sync to cloud services, with specific
spuport for GCP. Adding others -- so long as they support a
compare-and-set operation -- should be comparatively straightforward.

The cloud support includes cleanup of unnecessary data, and should keep
total space usage roughly proportional to the number of tasks.

Co-authored-by: ryneeverett <ryneeverett@gmail.com>

dependabot[bot] committed:
Bump webpki from 0.22.0 to 0.22.4 (#3246)

Bumps [webpki](https://github.com/briansmith/webpki) from 0.22.0 to 0.22.4.
- [Commits](https://github.com/briansmith/webpki/commits)

---
updated-dependencies:
- dependency-name: webpki
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

djmitche committed:
Update to supported linux distros (#3248)

djmitche committed:
Removed unused import (#3245)

Co-authored-by: Akash Shanmugaraj <65720968+akashShanmugraj@users.noreply.github.com>

ryneeverett committed:
sync: Use 600,000 iterations for pbkdf2. (#3251)

Per [wikipedia](https://en.wikipedia.org/wiki/PBKDF2), "In 2023, OWASP
recommended to use 600,000 iterations for PBKDF2-HMAC-SHA256..."

djmitche committed:
Better links from TW devel docs to TaskChampion (#3253)

ryneeverett committed:
sync server: Don't hash client_id for salt (#3250)

We don't know why we're doing this step so we probably shouldn't.
Cryptography isn't magic and extra steps are harmful in that they
obscure the important parts.

djmitche committed:
Rename `sync.server.encryption_secret` to `sync.encryption_secret`. (#3249)

* Rename `sync.server.encryption_secret` to `sync.encryption_secret`.

* move pwgen

djmitche committed:
Change 'should' to 'may' in server snapshot requirements (#3256)

For an object store backend, none of the suggestions in this paragraph
make much sense. "May" is much softer, in terms of
https://datatracker.ietf.org/doc/html/rfc2119. And in particular, the
cloud server implementation does not do any of these things.

djmitche committed:
Add note about Rust not building automatically (#3255)

dependabot[bot] committed:
Bump actions/cache from 3 to 4 (#3257)

Bumps [actions/cache](https://github.com/actions/cache) from 3 to 4.
- [Release notes](https://github.com/actions/cache/releases)
- [Changelog](https://github.com/actions/cache/blob/main/RELEASES.md)
- [Commits](https://github.com/actions/cache/compare/v3...v4)

---
updated-dependencies:
- dependency-name: actions/cache
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

akashShanmugraj committed:
fixed gcp service account key docs

akashShanmugraj committed:
refractor variable name

akashShanmugraj committed:
variable name refractor and datatype change

akashShanmugraj committed:
datatype change

akashShanmugraj committed:
`cargo fmt` changes

akashShanmugraj commented:
@djmitche I guess the datatype changes here are correct?
[a5ee2e8](https://github.com/akashShanmugraj/taskwarrior/commit/a5ee2e8e9c1549db20cf2c1605112f8b5afd631f) and [227cfb2](https://github.com/akashShanmugraj/taskwarrior/commit/227cfb2d85f20328c0fe2285f6117de34ff7e616)

But I feel that these commits lack _Code Quality_ ?

djmitche committed:
Use the correct formatting for `tag_` properties (#3261)

This resolves a discrepancy between Taskwarrior and Taskchampion as to
how tags should be represented.

djmitche commented on taskchampion/taskchampion/src/server/config.rs:
```diff
@@ -27,6 +31,17 @@ pub enum ServerConfig {
+        /// Bucket in which to store the task data. This bucket must not be used for any other
+        /// purpose.
+        bucket: String,
+        credential_path: Option<String>,
```
Looks good -- please add a docstring!

akashShanmugraj replied:
Okay!

djmitche commented on taskchampion/lib/src/server.rs:
````diff
@@ -154,6 +154,69 @@ pub unsafe extern "C" fn tc_server_new_remote(
+///
+/// ```c
+/// EXTERN_C struct TCServer *tc_server_new_gcp(struct TCString bucket,
+///                                       struct TCString credentia_path,
````
```suggestion
///                                       struct TCString credential_path,
```

djmitche commented on taskchampion/lib/src/server.rs:
```diff
@@ -154,6 +154,69 @@ pub unsafe extern "C" fn tc_server_new_remote(
+            //  - credential_path is valid (promised by caller)
+            //  - credential_path ownership is transferred to this function
+
+            let credential_path: String;
+
+            if credential_path_argument.is_null() {
+                credential_path = String::from("");
+            } else {
+                credential_path =
+                    unsafe { TCString::val_from_arg(credential_path_argument) }.into_string()?;
+            }
```
I talked about `is_null` earlier but I think this isn't actually what we want: TCStrings aren't actually allowed to be null. The `is_null` is a bit of a hack to handle some corner cases where there's a failure converting the TCString from one variant to the other. Specifically, there's no `tc_string_null` to create a null TCString from C.

Instead, I think we'll need to convert this to a `String`, and then check whether that is `""`.
```suggestion
            let credential_path =
                    unsafe { TCString::val_from_arg(credential_path_argument) }.into_string()?;
            let credential_path = if credential_path == "" {
                None
            } else {
                Some(credential_path)
            };
```

akashShanmugraj replied:
Okay!

djmitche reviewed (commented):
This is looking good!

The PR this is based on has now landed, so you can now make a PR against the https://github.com/GothenburgBitFactory/taskwarrior repository containing just your changes. That should be easier to review!

Also, please make sure to run `cargo fmt` and `cargo check` on the result.

akashShanmugraj committed:
Update typo suggestion

Co-authored-by: Dustin J. Mitchell <dustin@v.igoro.us>

akashShanmugraj committed:
Update datatype check

Co-authored-by: Dustin J. Mitchell <dustin@v.igoro.us>

akashShanmugraj committed:
Refactor credential_path assignment in tc_server_new_gcp function

akashShanmugraj commented on taskchampion/lib/src/server.rs:
```diff
@@ -202,7 +202,7 @@ pub unsafe extern "C" fn tc_server_new_gcp(
             let server_config = ServerConfig::Gcp {
                 bucket,
-                credential_path: Some(credential_path),
+                credential_path: credential_path.map(|cp| cp.to_string()),
```
@djmitche I got a _Expected String, got Option<String>_ (probably because of None) so I changed it like this
Is this commit fine?

djmitche replied:
It doesn't look right. What's the type of credential_path?

akashShanmugraj replied:
`credential_path` is a String, `credential_path_argument` is a TCString

djmitche replied:
Hm, looking up in this file I see 
```rust
let credential_path = if credential_path == "" {
                None
            } else {
                Some(credential_path)
            };
```
so that's a vlaue of type `Option<String>`. Don't forget that Rust allows new declarations of a variable name with a different type!

akashShanmugraj commented:
Okay, I will open a new PR to the develop branch

djmitche committed:
Fix invalid imports in `[{..}]` form (#3241)

* Fix invalid imports in `[{..}]` form

Before this change, if an import of data that takes the form of a JSON
array of JSON objects results in an error, the import would be re-tried
assuming that each line of the file is a JSON object (the old format).
However, no check was made that the value actually was an object before
casting it to `json::object`, resulting in a segfault.

This adds the check, and handles the failure with a useful error message
(from the first attempt to parse the file).

akashShanmugraj committed:
corrected documentation

akashShanmugraj committed:
modifed for Clippy Check

djmitche commented:
Now at https://github.com/GothenburgBitFactory/taskwarrior/pull/3262
